File: app/src/main/java/com/mxg/settings/module/hook/camera/CustomWatermark.java
// this file is a part of MxGSettings. if you kang it w/o permission, or you don't credit, you're COOKED. T-T _/\_
package com.mxg.settings.module.hook.camera;

import android.util.SparseArray;

import com.mxg.settings.module.base.BaseHook;
import com.mxg.settings.module.base.dexkit.DexKit;
import com.mxg.settings.module.base.dexkit.IDexKitList;

import org.luckypray.dexkit.DexKitBridge;
import org.luckypray.dexkit.query.FindMethod;
import org.luckypray.dexkit.query.matchers.AnnotationMatcher;
import org.luckypray.dexkit.query.matchers.AnnotationsMatcher;
import org.luckypray.dexkit.query.matchers.MethodMatcher;
import org.luckypray.dexkit.result.MethodDataList;

import java.lang.reflect.AnnotatedElement;
import java.lang.reflect.Method;
import java.util.List;

public class CustomWatermark extends BaseHook {
    @Override
    public void init() throws NoSuchMethodException {
        List<Method> methods = DexKit.getDexKitBridgeList("Watermark", new IDexKitList() {
            @Override
            public List<AnnotatedElement> dexkit(DexKitBridge bridge) throws ReflectiveOperationException {
                MethodDataList methodData = bridge.findMethod(FindMethod.create()
                        .matcher(MethodMatcher.create()
                                .returnType(SparseArray.class)
                                .paramCount(0)
                                .annotations(AnnotationsMatcher.create()
                                        .add(AnnotationMatcher.create()
                                                .usingStrings("Ljava/lang/String;")
                                        )
                                )
                        )
                );
                return DexKit.toElementList(methodData);
            }
        }).toMethodList();
        for (Method method : methods) {
            // Method method = methodData.getMethodInstance(lpparam.classLoader);
            logD(TAG, lpparam.packageName, "Current hooking method is " + method);
            hookMethod(method, new MethodHook() {
                @Override
                protected void before(MethodHookParam param) throws Throwable {
                    SparseArray<String[]> sparseArray = new SparseArray<>(1);
                    sparseArray.put(0, new String[]{mPrefsMap.getString("camera_custom_watermark_manufacturer", "XIAOMI"), mPrefsMap.getString("camera_custom_watermark_device", "MI PHONE")});
                    param.setResult(sparseArray);
                }
            });
        }
    }
}

File: app/src/main/java/com/mxg/settings/module/hook/mediaeditor/CustomWatermark.kt
// this file is a part of MxGSettings. if you kang it w/o permission, or you don't credit, you're COOKED. T-T _/\_
package com.mxg.settings.module.hook.mediaeditor

import com.github.kyuubiran.ezxhelper.*
import com.github.kyuubiran.ezxhelper.HookFactory.`-Static`.createHook
import com.mxg.settings.module.base.*
import com.mxg.settings.module.base.dexkit.*
import com.mxg.settings.module.base.dexkit.DexKitTool.toMethod
import org.luckypray.dexkit.query.enums.*

object CustomWatermark : BaseHook() {
    private val name by lazy {
        mPrefsMap.getString("mediaeditor_custom_watermark", "")
    }

    // by StarVoyager
    private val search by lazy {
        DexKit.getDexKitBridge("CustomWatermark") {
            it.findMethod {
                matcher {
                    addUsingString("K30 Pro Zoom E", StringMatchType.Equals)
                    // modifiers = Modifier.FINAL // 1.6.5.10.2 改成了 STATIC，所以寄了
                    returnType = "java.lang.String"
                    paramCount = 2
                }
            }.single().getMethodInstance(EzXHelper.classLoader)
        }.toMethod()
    }

    override fun init() {
        logD(TAG, lpparam.packageName, "[CustomWatermark] search method is $search")
        search.createHook {
            // 当前只能修改后缀
            returnConstant(name)
        }

        /*SystemProperties.methodFinder()
            .filterByParamCount(2)
            .filterByParamTypes(String::class.java, String::class.java)
            .toList().createHooks {
                before {
                    if (it.args[0] == "ro.product.marketname") {
                        it.args[1] = name
                    }
                }
            }

        SystemProperties.methodFinder()
            .filterByName("get")
            .filterByParamTypes(String::class.java)
            .toList().createHooks {
                before {
                    if (it.args[0] == "ro.product.marketname") {
                        it.result = name
                    }
                }
            }*/
    }
}
